File: confirm.go
/*
Copyright 2016 Nymi Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math/big"
)

// The following function handles request to verify a signature produced by the Nymi
// as a response to a roaming authentication sign. Typically, this will be integrated
// into another single sign-on system as a plugin to securely identify the wearer of the band.
func doConfirm(decoder *json.Decoder) (respJson []byte) {
	Debug.Println("confirm")
	var resp ConfirmationResponse

	var req ConfirmationRequest
	err := decoder.Decode(&req)
	if err != nil {
		resp.Okay = false
		resp.ErrorMsg = "failed to parse confirmation request"
		respJson, _ = json.MarshalIndent(resp, "", "  ")
		Info.Println("failed request:", resp.ErrorMsg)
		return
	}

	resp.NymiBandTID = req.NymiBandTID

	serverNonce, err := hex.DecodeString(req.ServerNonce)
	if len(serverNonce) != 32 {
		resp.Okay = false
		resp.ErrorMsg = fmt.Sprintf("server-nonce has incorrect length of %d, must be 32 bytes.", len(serverNonce))
		respJson, _ = json.MarshalIndent(resp, "", "  ")
		Info.Println("failed request:", resp.ErrorMsg)
		return
	}

	signature, err := hex.DecodeString(req.NymiBandSignature)
	if len(signature) != 64 {
		resp.Okay = false
		resp.ErrorMsg = fmt.Sprintf("signature has incorrect length of %d, must be 64 bytes.", len(signature))
		respJson, _ = json.MarshalIndent(resp, "", "  ")
		Info.Println("failed request:", resp.ErrorMsg)
		return
	}

	user, ok := users[req.VerificationKeyId]
	if !ok {
		resp.Okay = false
		resp.ErrorMsg = "unknown verification-key-id."
		respJson, _ = json.MarshalIndent(resp, "", "  ")
		Info.Println("failed request:", resp.ErrorMsg)
		return
	}

	// The value of req.ConfirmingNEA should be checked here

	inputHash := sha256.Sum256(serverNonce)

	var pubkey ecdsa.PublicKey
	pubkey.Curve = elliptic.P256()
	vkBytes := user.VerificationKey.Bytes()
	pubkey.X = big.NewInt(0).SetBytes(vkBytes[0:32])
	pubkey.Y = big.NewInt(0).SetBytes(vkBytes[32:64])

	inr := big.NewInt(0).SetBytes(signature[0:32])
	ins := big.NewInt(0).SetBytes(signature[32:64])

	// Verify the signature
	verificationResult := ecdsa.Verify(&pubkey, inputHash[:], inr, ins)
	if verificationResult {
		resp.Okay = true
		resp.UserData = user.Data
		resp.ConfirmingNEA = req.ConfirmingNEA
	} else {
		resp.Okay = false
		resp.ErrorMsg = "failed to verify signature"
	}

	respJson, _ = json.MarshalIndent(resp, "", "  ")
	return
}

File: lib/confirm.js
'use strict';

var Config   = global.Config,
    Log      = global.Log,
    util     = require('util'),
    crypto   = require('crypto'),
    EC       = require('elliptic').ec,
    express  = require('express'),
    camelize = require('camelize'),
    json     = require('jsonfile2'),
    Server   = require('./server.js'),
    Confirm  = express.Router(),
    ReqProps = ['tid', 'server-nonce', 'verification-key-id', 'nymiband-signature', 'confirming-nea'],
    ResProps = ['okay', 'error', 'tid', 'user-data', 'confirming-nea']
;

Confirm.use(Server.addCORSHeader);

Confirm.doConfirm = function(req)
{
    Log.debug('confirm');

    var users, sha256, inputHash, ec, user, signature, verificationResult;

    if (!util.isObject(req.body) || !ReqProps.every(req.body.hasOwnProperty, req.body)) {
        return Server.marshalResponse(ResProps, {
            okay: false,
            error: 'failed to parse confirmation request'
        });
    }

    req = camelize(req.body);

    if (req.serverNonce.length != 64) {
        return Server.marshalResponse(ResProps, {
            okay: false,
            error: util.format('server-nonce has incorrect length of %d, must be 32 bytes.', req.serverNonce.length >> 1)
        });
    }

    if (req.nymibandSignature.length != 128) {
        return Server.marshalResponse(ResProps, {
            okay: false,
            error: util.format('signature has incorrect length of %d, must be 64 bytes.', req.nymibandSignature.length)
        });
    }

    users = new json.File(Config.get('UsersFileName'));
    users.readSync();

    if (users.get(req.verificationKeyId) === undefined) {
        return Server.marshalResponse(ResProps, {
            okay: false,
            error: 'unknown verification-key-id.'
        });
    }

    user = users.get(req.verificationKeyId);

	// The value of req.confirmingNea should be checked here

    sha256 = crypto.createHash('sha256');
    sha256.update(new Buffer(req.serverNonce, 'hex'));

    inputHash = sha256.digest('hex');
    signature = {r: req.nymibandSignature.slice(0, 64), s: req.nymibandSignature.slice(64)};

    ec = new EC('p256');
    verificationResult = ec.verify(inputHash, signature, 0x04 + user.VerificationKey, 'hex');

    if (!verificationResult) {
        return Server.marshalResponse(ResProps, {
            'okay':  false,
            'tid':   req.tid,
            'error': 'failed to verify signature'
        });
    }

    return Server.marshalResponse(ResProps, {
        'okay':           true,
        'tid':            req.tid,
        'user-data':      user.Data,
        'confirming-nea': req.confirmingNea
    });
};

Confirm.options('/', function (req, res, next) {
    Log.debug('preflight /confirm');

    res.status(204).send();
});

Confirm.post('/', function (req, res, next) {
    Log.debug('post /confirm');

    return res.json(Confirm.doConfirm(req));
});

module.exports = Confirm;